dask/blockwise.py

```diff
@@ -1355,9 +1357,17 @@ def rewrite_blockwise(inputs):
 
             changed = True
 
+            # Check if this dep was already encountered.
+            # If so, we should append a "counter" label
+            # to the name to avoid fusing distinct indices
+            # into a single dependency within the subgraph
+            # (see: https://github.com/dask/dask/issues/8535)
+            use_dep = dep + f"_{seen[dep]}" if seen[dep] else dep
+            seen[dep] += 1
```

Comment:
```suggestion
            use_dep = f"{dep}_{''.join(ind)}"
```
Still wondering whether we actually need the `seen` set, or whether we can just include the indices in the name in every case. As far as I understand, the key names in this `dsk` are an implementation detail of Blockwise and could be changed to whatever as long as the dask is still valid (is this correct??).

Making this change on your branch makes your `test_optimize_blockwise_duplicate_dependency` pass, along will all other Array tests _besides_ `dask/array/tests/test_atop.py::test_rewrite`. The reason `test_rewrite` fails is that its cases include handwritten dasks of the expected output, so obviously the key names there no longer match. However, I don't know if those key names matter from a correctness standpoint, or if we could just adjust those tests. (The fact that they're not failing with this branch just indicates to me that none of those cases involve this situation of a repeated dep, so they're incomplete anyway.)

I'd guess the only case where the key names matter is with `output`, which (as I mentioned in the other comment) probably needs to be addressed here too.

Reply:
Right - We could certainly add an enumeration in every key that is added to the root subgraph during fusion. This should result in "correct" behavior, and changing the tests should be fine in my opinion. With that said, this approach wont really simplify the code much.

Reply:
In this very complicated code, I'll take any simplification I can get :)

It also is slightly different behavior, and I'm not sure which one we want. In my version, two repeated dep names with the same (user-given?) indices will still trample each other. Maybe this is good, because they should be considered identical? Or maybe we want to consider every layer unique, even if the names and indices match. I'm really not sure.